### ISSUE TITLE
Timeout::timeout not throwing exception when timeout period reaches

### ISSUE THREAD

**Issue by navaneeth** [jruby/jruby#1050]
_/tmp/loop.sh_

``` shell
#!/bin/sh
while :
do
    sleep 1
done
```

Now start the above script from a Ruby script.

_/tmp/test.rb_

``` ruby
require 'timeout'
begin
    Timeout::timeout(1) {
        `/tmp/loop.sh`.chomp
    }
rescue Exception => e
    puts 'inside rescue'
rescue Error => e1
    puts "inside error"
end
```

Execute the above script in a JRuby `ScriptingContainer`.

``` java
ScriptingContainer container = new ScriptingContainer();
container.runScriptlet(PathType.ABSOLUTE, "/tmp/test.rb");
```

The rescue block is not getting executed. Tested this with MRI and it works well there.

**Comment by rtyler** [comment]
FWIW, I've reproduced that this is still the case with JRuby 9.0.0.0 whereas MRI 2.1.5 puts "inside rescue"

**Comment by Adithya-copart** [comment]
I faced the same issue when I tried to do something like this on :

```
require 'timeout'
begin
  Timeout.timeout(5) do
    pid = spawn('ruby -e "sleep"')
    Process.wait pid
  end
rescue Timeout::Error => e
  'timed out'
end
```

**Comment by headius** [comment]
I suspect MRI is using a non-blocking form of wait here, that allows them to give up on the subprocess. I will look into whether we can do the same.

**Comment by headius** [comment]
Fixed by #5848.
